exe/proteus

```diff
@@ -74,79 +81,252 @@ command :info do |c|
+            rescue Proteus::ApiEntityError::EntityNotFound => e
+              puts "#{name} not found in zone #{hint} with id #{z.id}"
+            end
+          end
+        end
+      end
+    end
+  end
+end
+
+#####
+# Actions specific to external (unmanaged) records
+desc 'Performs actions on external records'
+command :external do |e|
+  e.desc 'Search for an external record'
+  e.arg_name 'keyword'
+  e.command :search do |s|
+    s.action do |_global_options, _options, args|
+      validate_argnum(1, args.length)
+      ap @proteus.search_by_object_types(args[0], Proteus::Types::EXTERNALHOST)
+    end
+  end
+
+  e.desc 'Add external record'
+  e.arg_name 'name'
+  e.command :add do |a|
     a.action do |_global_options, _options, args|
+      validate_argnum(1, args.length)
+      ap @proteus.add_external_host_record(args[0])
+    end
+  end
+end
+
+#####
+# Actions specific to IPv4 address management
+desc 'Manage IPv4 addresses'
+command :ip do |i|
+  i.desc 'Identify the network by an existing ip in the network'
+  i.switch [:i, :useip], negatable: false
+
+  i.desc 'Show an IP address'
+  i.arg_name 'ipaddress [configuration_id]'
+  i.command :show do |s|
+    s.action do |_global_options, _options, args|
+      raise "command requires 1 args" unless args.length >= 1
+      config_id = args[1] || @proteus.get_entities.first.id
+      puts @proteus.get_ip4_address(config_id, args[0])
+    end
+  end
+
+  i.desc 'Show the next available IP address in the network.'
+  i.arg_name 'cidr|ipaddress'
+  i.command :next do |n|
+    n.action do |_global_options, options, args|
+      validate_argnum(1, args.length)
+      config_id = @proteus.get_entities.first.id
+
+      # If we are passed an IP, use it.  Otherwise generate a canary.
+      # Note: it's _possible_ the first IP in a net block is unallocated, so
+      # there could be an edge case where this fails.
+      if options[:useip] || args[0].end_with?('/32')
+        canary_ip = NetAddr::CIDR.create(args[0]).first
+      else
+        # first assignable IP (different from the _first_ IP)
+        canary_ip =  NetAddr::CIDR.create(args[0]).nth(1)
+      end
+
+      ipv4_id = @proteus.get_ip4_address(config_id, canary_ip).id
+      parent_id = @proteus.get_parent(ipv4_id).id
+      puts @proteus.get_next_ip4_address(parent_id)
+    end
+  end
+
+  i.desc 'Assigns the next available IP address in the network.'
+  i.arg_name 'cidr|ipaddress fqdn'
+  i.command :assign do |a|
+    a.action do |_global_options, options, args|
       validate_argnum(2, args.length)
-      ap @proteus.add_host_record(args[0], args[1])
+      config_id = @proteus.get_entities.first.id
+
+      # If we are passed an IP, use it.  Otherwise generate a canary.
+      # Note: it's _possible_ the first IP in a net block is unallocated, so
+      # there could be an edge case where this fails.
+      if options[:useip] || args[0].end_with?('/32')
+        canary_ip = NetAddr::CIDR.create(args[0]).first
```

Comment:
In this case I don't think so -- 'network' implies `x.x.x.0` for example, but this is a `/32`